# Assignment 1
You should submit the **UniversityNumber.ipynb** file and your final prediction file **UniversityNumber.test.txt** to moodle. Make sure your code does not use your local files and that the results are reproducible. Before submitting, please **1. clean all outputs and 2. run all cells in your notebook and keep all running logs** so that we can check.

# 1 $n$-gram Language Model

In [2]:
!mkdir -p data/lm
!wget -O data/lm/train.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/lm/train.txt
!wget -O data/lm/dev.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/lm/dev.txt
!wget -O data/lm/test.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/lm/test.txt

The syntax of the command is incorrect.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.


## 1.1 Building vocabulary

### Code

In [56]:
with open('data/lm/train.txt', 'r', encoding="utf-8") as f:
    train = dict()
    for line in f:
        line = '<START> ' + line + ' <END>'     # to add <START> and <END> to the start and end of each sentence 
        tmp = line.split()
        for word in tmp:
            if word in train.keys():            # if in dict, increment counter
                train[word] += 1
            else:
                train[word] = 1                 # if not in dict, add new entry

    countUNK = 0
    for key in list(train.keys()):              # find all words with freq < 3, sum their counts
        if train[key] < 3:
            countUNK += train[key]
            del train[key]
    train['<UNK>'] = countUNK
    print("Size of vocabulary = " + str(len(train)))

Size of vocabulary = 26601


### Discussion

Size of vocabulary = 26601

Number of parameters of n-gram models = (size of vocabulary) ^ n = (26601) ^ n
eg unigram: 26601 ^ 1
eg bigram: 26601 ^ 2
eg trigram: 26601 ^ 3

## 1.2 $n$-gram Language Modeling

### Code

In [70]:
import math

unigramDict = dict()
bigramDict = dict()

# build unigram
def unigram():
    for key in train.keys():
        unigramDict[key] = float(train[key]) / sum(train.values())

def unigramProb(word):
    if word in unigramDict.keys():
        return unigramDict[word]
    else:
        return unigramDict['<UNK>']

# build bigram
def bigram():
    with open('data/lm/train.txt', 'r', encoding="utf-8") as f:
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            for i in range(len(tmp) - 1):
                if tmp[i] not in train.keys():
                    tmp[i] = '<UNK>'
                if tmp[i+1] not in train.keys():
                    tmp[i+1] = '<UNK>'
                    
                if (tmp[i], tmp[i+1]) in bigramDict.keys():
                    bigramDict[(tmp[i], tmp[i+1])] += 1
                else:
                    bigramDict[(tmp[i], tmp[i+1])] = 1
    for key in bigramDict.keys():
        # bigramDict[key] = float(bigramDict[key])     # P(w2|w1) = C(w1, w2) / C(w1)
        bigramDict[key] = float(bigramDict[key]) / train[key[0]]    # P(w2|w1) = C(w1, w2) / C(w1)
        # bigramDict[key] = train[key[0]]    # P(w2|w1) = C(w1, w2) / C(w1)
                    
unigram()
# bigram()
print(train['the'])
print(unigramDict['the'])
print(sum(train.values()))
print(str(len(train)))
print(train['exceeding'])
print(unigramDict['exceeding'])
print("Unigram probability of 'the' = " + str(unigramDict['the']))
# print("Bigram probability of 'the' given 'the' = " + str(bigramDict[('the', 'the')]))


72045
0.04293396708783213
1678042
26601
9
5.3633937648759684e-06
Unigram probability of 'the' = 0.04293396708783213


In [71]:
def unigramTrainPerplexity():
    with open('data/lm/train.txt', 'r', encoding="utf-8") as f:
        N = 0
        sum = 0
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            print(tmp)
            prob = 1
            for word in tmp:
                N += 1
                prob += math.log2(unigramProb(word))
                print(prob, unigramProb(word), word, line)
            sum += math.log2(prob)
    return 2 ** (-sum / N)
    
def unigramDevPerplexity():
    with open('data/lm/dev.txt', 'r', encoding="utf-8") as f:
        perplexity = 1
        N = 0
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            for word in tmp:
                N += 1
                perplexity *= 1 / unigramProb(word)
        perplexity = math.pow(perplexity, 1/float(N))
        return perplexity

def bigramTrainPerplexity():
    with open('data/lm/train.txt', 'r', encoding="utf-8") as f:
        perplexity = 1
        N = 0
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            for i in range(len(tmp) - 1):
                N += 1
                if (tmp[i], tmp[i+1]) in bigramDict.keys():
                    perplexity *= 1 / bigramDict[(tmp[i], tmp[i+1])]
                else:
                    perplexity *= 1 / unigramProb(tmp[i+1])
        perplexity = math.pow(perplexity, 1/float(N))
        return perplexity
    
def bigramDevPerplexity():
    with open('data/lm/dev.txt', 'r', encoding="utf-8") as f:
        perplexity = 1
        N = 0
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            for i in range(len(tmp) - 1):
                N += 1
                if (tmp[i], tmp[i+1]) in bigramDict.keys():
                    perplexity *= 1 / bigramDict[(tmp[i], tmp[i+1])]
                else:
                    perplexity *= 1 / unigramProb(tmp[i+1])
        perplexity = math.pow(perplexity, 1/float(N))
        return perplexity
    
print("Unigram train perplexity = " + str(unigramTrainPerplexity()))


['<START>', 'Newquay', "'s", 'senior', 'police', 'officer', ',', 'Insp', 'Dave', 'Meredith', ',', 'said', 'the', 'application', 'would', 'be', 'looked', 'at', 'closely', '.', '<END>']
-3.774842084590672 0.0365282871346486 <START> <START> Newquay 's senior police officer , Insp Dave Meredith , said the application would be looked at closely .
 <END>
-22.45318947960522 2.383730562167097e-06 Newquay <START> Newquay 's senior police officer , Insp Dave Meredith , said the application would be looked at closely .
 <END>
-29.38166744722293 0.008209568056103482 's <START> Newquay 's senior police officer , Insp Dave Meredith , said the application would be looked at closely .
 <END>
-42.12337690323491 0.0001460034969327347 senior <START> Newquay 's senior police officer , Insp Dave Meredith , said the application would be looked at closely .
 <END>
-52.93444555853981 0.0005566010862660172 police <START> Newquay 's senior police officer , Insp Dave Meredith , said the application would be look

ValueError: math domain error

### Discussion

## 1.3 Smoothing

### 1.3.1 Add-one (Laplace) smoothing

### Code

#### Discussion

#### 1.3.2 Add-k smoothing

##### Code

##### Discussion

### 1.3.3 Linear Interpolation

#### Code

#### Discussion

##### 1.3.4 Optimization

#### Discussion

# 2 Word Vectors

In [1]:
def load_embedding_model():
    """ Load GloVe Vectors
        Return:
            wv_from_bin: All 400000 embeddings, each lengh 50
    """
    import gensim.downloader as api
    wv_from_bin = api.load("glove-wiki-gigaword-200")
    print("Loaded vocab size %i" % len(list(wv_from_bin.index_to_key)))
    return wv_from_bin

wv_from_bin = load_embedding_model()

[==================================================] 100.0% 252.1/252.1MB downloaded
Loaded vocab size 400000


## 2.1 Find most similar word
Use cosine similarity to find the most similar word to each of these words. Report the most similar word and its cosine similarity.

In [2]:
import numpy as np
from numpy.linalg import norm

wordList = ["dog", "whale", "before", "however", "fabricate"]
wordDict = {}
vecKeys = wv_from_bin.index_to_key
for word in wordList:
  sim = -1
  most = ""
  for i in range(len(wv_from_bin)):
    x = wv_from_bin[word]
    cosine = np.dot(x,wv_from_bin[i])/(norm(x)*norm(wv_from_bin[i]))
    curword = vecKeys[i]
    if cosine >= sim and curword != word:
      sim = cosine
      most = curword
  wordDict[word] = [most, sim]
print("Answer format: {word: [most similar word, cosine similarity]}")
print(wordDict)

Answer format: {word: [most similar word, cosine similarity]}
{'dog': ['dogs', 0.81368625], 'whale': ['whales', 0.7918058], 'before': ['after', 0.8931248], 'however': ['although', 0.9336163], 'fabricate': ['fabricating', 0.61840194]}


## 2.2 Finding Analogies
Use vector addition and subtraction to compute target vectors for the analogies below. After computing each target vector, find the top three candidates by cosine similarity. Report the candidates and their similarities to the target vector.

- dog : puppy :: cat : ?:
- speak : speaker :: sing : ?:
- France : French :: England : ?:
- France : wine :: England : ?

In [5]:
# dog-puppy+cat
wordList = [["dog", "puppy", "cat"], ["speak", "speaker", "sing"], ["france", "french", "england"], ["france", "wine", "england"]]

vecKeys = wv_from_bin.index_to_key
for analogy in wordList:
  targetVec = wv_from_bin[analogy[0]] - wv_from_bin[analogy[1]] + wv_from_bin[analogy[2]]
  
  candidateList = []
  similarityList = []
  for i in range(len(wv_from_bin)):
    cosine = np.dot(targetVec,wv_from_bin[i])/(norm(targetVec)*norm(wv_from_bin[i]))
    curword = vecKeys[i]
    
    if len(similarityList) < 3 or (len(similarityList) >= 3 and cosine >= similarityList[2]):
      similarityList.append(cosine) 
      similarityList.sort(reverse=True)
      candidateList.insert(similarityList.index(cosine), curword)
  print(f"{analogy[0]} : {analogy[1]} :: {analogy[2]} : ?:", candidateList[0:3])
  print("corresponding similarities:", similarityList[0:3])
# targetVec = wv_from_bin["dog"] - wv_from_bin["puppy"] + wv_from_bin["cat"]
# print("kitty", np.dot(targetVec,wv_from_bin["kitty"])/(norm(targetVec)*norm(wv_from_bin["kitty"])))
# print("kitten", np.dot(targetVec,wv_from_bin["kitten"])/(norm(targetVec)*norm(wv_from_bin["kitten"])))
# print("dogs", np.dot(targetVec,wv_from_bin["dogs"])/(norm(targetVec)*norm(wv_from_bin["dogs"])))

dog : puppy :: cat : ?: ['cat', 'dog', 'dogs']
corresponding similarities: [0.8246017, 0.745641, 0.6278181]
speak : speaker :: sing : ?: ['sing', 'speak', 'singing']
corresponding similarities: [0.7365669, 0.55140996, 0.548521]
france : french :: england : ?: ['england', 'wales', 'ireland']
corresponding similarities: [0.837254, 0.6498687, 0.6262423]
france : wine :: england : ?: ['england', 'france', 'britain']
corresponding similarities: [0.69291466, 0.6897908, 0.61666137]


# 3 Sentiment analysis

In [ ]:
!mkdir -p data/classification
!wget -O data/classification/train.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/classification/train.txt
!wget -O data/classification/dev.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/classification/dev.txt
!wget -O data/classification/test-blind.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/classification/test-blind.txt

The syntax of the command is incorrect.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.


## 3.1 Logistic Regression

### Unigram Features

### Bigram Features

### GloVe Features

Compare the performance of three types of features on dev set. Report the weighted average precision, recall and F1-score for each feature set.

| Feature | precision | recall | F1-score |
| ----------- | --------- | ------ | -------- |
| unigram     |           |        |          |
| bigram      |           |        |          |
| GloVe       |           |        |          |

## 3.2 Better Feature